solveset : nonlinsolve - nonlinear system of equation

Shekharrajak opened the pull request:
Non linear system of equation solver.
- [x] Modify testcases.
- [x] Non poly cases ( where `_solve_poly_system` can't help) .
- [x] Import ALL the testcases from old `solve` and add new testcases.
- [x] More clear comments and docstring.
- [ ] Address https://github.com/sympy/sympy/pull/11111#issuecomment-226851817

Shekharrajak committed:
testcase for nonlinear system of eq from old solve

Shekharrajak committed:
nlinsolve created with initial code

minor edits in test

Shekharrajak commented:
ping @hargup @certik @aktech @kshitij10496 @asmeurer .
Please check this PR . Work is in progress.

Shekharrajak committed:
created nlinsolve_matrix

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1118,154 @@ def linsolve(system, *symbols):
+    >>> nlinsolve((e1, e2), (x, y))
+    >>> {(191/20, -3*sqrt(391)/20), (191/20, 3*sqrt(391)/20)}
+
+    """
```
The docstrings can be improved.
Here are a few suggestions from my side:
- a bit more explanation about the method (API) can help
- few more examples (different types of system) would look cool.

A reference to `linsolve` might come in handy.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1118,154 @@ def linsolve(system, *symbols):
+                         'system is to be found.')
+    polys = []
+    failed = []
+    for j, g in enumerate(system):
```
I think `enumerate` might not be an overkill here.

kshitij10496 replied:
A simpler iteration technique can be used here.

Shekharrajak replied:
Thanks, I'll modify it in next commit :+1:

hargup replied:
> I think `enumerate` might not be an overkill here.

Just curious, what do you mean by by an 'overkill'?

Shekharrajak commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1118,154 @@ def linsolve(system, *symbols):
+    This method will help in this type of cases(nlinsolve calling this method when 
+    solve_poly_system cannot solve non zero dimensional nonlinear system of equations) :
+    >>> nlinsolve([(x + y)**2 - 4, x + y - 2],[x,y])
+    >>> {-y + 2}
```
here output should be `x: -y+2` . If someone knows good way to do this then please suggest.

kshitij10496 replied:
I think the API should be compatible with that of `linsolve`.

With that thought in mind, the ideal output should be
`{(-y + 2, y)}`

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1118,154 @@ def linsolve(system, *symbols):
+
+def nlinsolve_matrix(system, symbols):
+    r"""
+    reference : http://people.math.gatech.edu/~aleykin3/math4803spr13/BOOK/chapter1.pdf
```
Please use the same documentation style as `linsolve`, first a good description of the method then illustrative examples. Also the references should go the reference section.

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1118,154 @@ def linsolve(system, *symbols):
+    return result
+
+
+def nlinsolve(system, symbols):
```
How is `nlinsolve` different from `nlinsolve_matrix`?

Shekharrajak replied:
`nlinsolve_matrix` can recognize if system of equation have infinite solutions( Matrix det becomes zero at that time).But now I am thinking about solve positive dimensional system using polynomial univariate representation (PUR) methid : 
* http://www.m-hikari.com/imf-2010/5-8-2010/ayadIMF5-8-2010.pdf
* http://research.cs.tamu.edu/keyser/Papers/AMSDimacs2005.pdf
* http://www.maplesoft.com/support/help/Maple/view.aspx?path=GroebneR/RationalUnivariateRepresentation

but not getting how to find `RUR` also called `PUR` .

hargup replied:
No, I mean I don't understand why there is a need for two methods `nlinsolve` and `nlinsolve_matrix`?

hargup commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1012,6 +1012,199 @@ def test_linsolve():
     assert linsolve(Augmatrix, a, b, c, d, e) == FiniteSet((a, 0, c, 0, e))
 
 
+# non linear system of equations
```
Are these tests imported from solve or have you written them yourself?

Shekharrajak replied:
Most of them are from `solve`.

Shekharrajak committed:
else part

Shekharrajak committed:
substitution function added.

substitution method modified

Shekharrajak committed:
description added

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,382 @@ def linsolve(system, *symbols):
+                                # solved symbols: discard it
+                                # eg : depend_soln=> {x : -y} and
+                                # r1 is function of x like r1 = -x for y
+                                skip = True
```
Is this line intended to be outside the `for` block ?

Shekharrajak replied:
No

Shekharrajak committed:
testcases, doctest  done

Shekharrajak commented:
My new commits are not coming here. Looks like github problem.

Shekharrajak committed:
All soln in finiteset

doctest edited

Shekharrajak commented:
```
In [1]: nlinsolve([exp(x) - sin(y), 1/exp(y) - 3], [x, y]) 
Out[1]: {{x: {ⅈ⋅(2⋅n⋅π + π) + log(sin(log(3))) | n ∊ ℤ}, y: -log(3)}}

In [2]: nlinsolve([exp(x) - sin(y), y**2 - 4], [x, y])
Out[2]: {{x: log(sin(2)), y: 2}, {x: {ⅈ⋅(2⋅n⋅π + π) + log(sin(2)) | n ∊ ℤ}, y: -2}}

In [3]: FiniteSet(Dict(\
   ...:         (x, imageset(Lambda(n, I*(2*n*pi + pi) + log(sin(log(3)))), S.Integers)),\
   ...:          (y, - log(3))))
Out[3]: {{x: {ⅈ⋅(2⋅π⋅n + π) + log(sin(log(3))) | n ∊ ℤ}, y: -log(3)}}

In [4]: FiniteSet(Dict((x, log(sin(S(2)))), (y, S(2))), \
   ...:         Dict((x, imageset(Lambda(n, I*(2*n*pi + pi) + log(sin(2))), S.Integers)),\
   ...:          (y, -2)))
Out[4]: {{x: log(sin(2)), y: 2}, {x: {ⅈ⋅(2⋅π⋅n + π) + log(sin(2)) | n ∊ ℤ}, y: -2}}
```
When we use SymPy defined dictionary `Dict` then we need to write testcase then we need to write testcases like this : 

```
assert nlinsolve([exp(x) - sin(y), 1/exp(y) - 3], [x, y]) == \
    FiniteSet(Dict(\
        (x, imageset(Lambda(n, I*(2*n*pi + pi) + log(sin(log(3)))), S.Integers)),\
         (y, - log(3))))

    assert nlinsolve([exp(x) - sin(y), y**2 - 4], [x, y]) == \
        FiniteSet(Dict((x, log(sin(S(2)))), (y, S(2))), \
        Dict((x, imageset(Lambda(n, I*(2*n*pi + pi) + log(sin(2))), S.Integers)),\
         (y, -2)))
```
But still these are failing [here](https://github.com/sympy/sympy/pull/11111/files#diff-85baa04bbf4e1dfd9128782738e45424R1054). I didn't understand why.

Shekharrajak committed:
testcases edited

Shekharrajak committed:
FiniteSet of ordered tuple for all soln

Shekharrajak commented:
Please have a look in this travis [error](https://travis-ci.org/sympy/sympy/jobs/133549546) . It works fine but don't know why assertion error.

Shekharrajak committed:
remove extra soln sqrt and denom

Shekharrajak commented:
@aktech I am still getting the [error](https://travis-ci.org/sympy/sympy/jobs/133549546) . I am trying to fix `test_issues_2777` but problem is solveset returns some answer instead if EmptySet so need to fix that first. I have added code to remove extra soln by checking denominator,sqrt expr.

Shekharrajak committed:
rename and comparison error fixed

Shekharrajak commented:
I have changed the name to `nonlinsolve` and fixed comparison error.

Shekharrajak committed:
Minor changes

Shekharrajak commented:
This PR is ready. Please have a look.

Shekharrajak commented:
Ping @aktech @hargup

aktech commented:
> @aktech I am still getting the error . I am trying to fix test_issues_2777 but problem is solveset returns some answer instead if EmptySet so need to fix that first. I have added code to remove extra soln by checking denominator,sqrt expr.

Is this fixed now?

Shekharrajak commented:
@aktech , changes for that is in [this commit](https://github.com/sympy/sympy/pull/11111/commits/598c0af1a44ded28a3b3007d0172e2ff55f51fd4). But [test_issue_2777](https://github.com/sympy/sympy/pull/11111/files#diff-85baa04bbf4e1dfd9128782738e45424R1114) is still in `XFAIL`, because of issue #11184 (solveset returned soln instead of Emptyset).

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+##############################################################################
+
+
+def substitution(system, symbols, result, known_symbols, exclude, all_symbols):
```
Does this needs to be available in public API?

Shekharrajak replied:
>Does this needs to be available in public API?

I didn't understand Public API ?

It is helper function for `nonlinsolve` but able to solve simple system.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    ==========
+
+    system : list of equations.
+
```
Remove empty lines

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+
+    system : list of equations.
+
+    symbols : list of unsolved symbols.
```
Use proper variable names. Explicit is better than implicit.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+
+    symbols : list of unsolved symbols.
+
+    known_symbols : list of solved symbols.
```
What do you mean by this? document properly.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+
+    known_symbols : list of solved symbols.
+
+    result : If non of the symbols are solved then it is empty list, otherwise
```
non --> none

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    known_symbols : list of solved symbols.
+
+    result : If non of the symbols are solved then it is empty list, otherwise
+    list of already solved_syms symbols( dict symbol : value )
```
space after symbols,
remove space after `(`

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    result : If non of the symbols are solved then it is empty list, otherwise
+    list of already solved_syms symbols( dict symbol : value )
+
+    exclude : Set of expression which should not be zero with result.
```
Please don't assume anything while documenting your code for the developer who would read this. It should make proper sense and write everything explicitly, what do you mean by the statement.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+
+    Returns
+    =======
+    Returns a FiniteSet of ordered tuple.
```
> Returns a FiniteSet of ordered tuple.

Of what?

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    ========
+
+    >>> from sympy.core.symbol import symbols
+    >>> x, y = symbols('x, y', real = True)
```
`real = True` --> `real=True` Please follow pep8.
Please fix pep8 Errors.
Tip: Use a pep8 plugin to check for pep8 errors for your code editor.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    >>> from sympy.core.symbol import symbols
+    >>> x, y = symbols('x, y', real = True)
+    >>> from sympy.solvers.solveset import substitution
+    >>> substitution([x +y], [x], [{y : 1}], [y], set([]), [x, y])
```
space around operator,
check pep8 of "How to write a dictionary."

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    >>> substitution([x +y], [x], [{y : 1}], [y], set([]), [x, y])
+    {(-1, 1)}
+    >>> substitution([x +y], [x], [{y : 1}], [y], set([x + 1]), [y, x])
+    EmptySet()
```
Why is this `EmptySet`?

Shekharrajak replied:
Here `exclude` expression is `set([x + 1])` so `x  = -1` should be discarded.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    from sympy import Complement
+    from sympy.core.containers import Dict
+
+    def _ok_syms(e, sort=False):
```
Use proper variable names, don't use names like `e`. Explicit is better than implicit.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+
+    def _ok_syms(e, sort=False):
+        # get the unsolved symbol.
+            rv = (e.free_symbols - set(known_symbols)) & set(all_symbols)
```
same here for `rv`

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+                rv.sort(key=default_sort_key)
+            return rv
+
+    # sort such that equation with the fewest potential symbols is first
```
What is a `potential symbol` ?

Shekharrajak replied:
Means equation having less unsolved symbol will be solved first. I have added this line in new commit.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+                    # do the same with soln_complex.
+                    # soln_complex = solveset_complex(eq2, s)
+                    if not soln:
+                        soln = solveset_complex(eq2, s)
```
What's the motivation behind solving in complex domain only when there is no solution in real domain?

Shekharrajak replied:
When I tried to add both complex and real solution. I found that complex solution comes complicated which makes the further steps hard. Also sometimes `solveset` returns same values for both real and complex domain.
So for now if no real solution then only using complex solution. (means either real or complex solution).

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+                    not_solvable = True
+                elif isinstance(soln, Complement):
+                    # extract solution and complement
+                    complements[s] = list(soln.args[1])[0]
```
What are you extracting from `Complement` here?
What's this `list(soln.args[1])[0]` mean?

```python
In [29]: soln = Complement(ImageSet(Lambda(x, x**2), N), Interval(0, oo))

In [30]: isinstance(soln, Complement)
Out[30]: True

In [31]: list(soln.args[1])[0]
..
TypeError: 'Interval' object is not iterable
```

Shekharrajak replied:
Thanks, updated.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+                elif isinstance(soln, Intersection):
+                    # sometimes solveset returns Intersection with S.Real
+                    # when #11174 is fixed,then this line should be removed
+                    soln = soln.args[1]
```
Same here:

```python
In [35]: soln = Intersection(ImageSet(Lambda(x, x**2), N), Interval(0, oo))

In [36]: isinstance(soln, Intersection)
Out[36]: True

In [37]: soln.args[1]
Out[37]: ImageSet(Lambda(x, x**2), Naturals())
```

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+
+def nonlinsolve(system, symbols):
+    r"""
+    Solve system of N non linear equations with M variables, which
```
Make full use of 79 characters in a line.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    r"""
+    Solve system of N non linear equations with M variables, which
+    means both under - and overdetermined systems are supported.
+    Positive dimensional system is also supported (Infinite solution).
```
Explain a bit about, what is positive dimensional system.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+        The target system of equations
+    symbols : list of Symbol
+
+    * Note : List of symbol is used here.
```
Why not an iterable in general?

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    >>> nonlinsolve([x*y - 1, 4*x**2 + y**2 - 5], [x, y])
+    {(-1, -1), (-1/2, -2), (1/2, 2), (1, 1)}
+
+
```
remove one line here

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    >>> from sympy.solvers.solveset import nonlinsolve
+    >>> x, y = symbols('x, y', real = True)
+    >>> nonlinsolve([x*y - 1, 4*x**2 + y**2 - 5], [x, y])
+    {(-1, -1), (-1/2, -2), (1/2, 2), (1, 1)}
```
All code blocks should be in Examples.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+
+    * If solveset (substitution method) returns complement
+    for any symbol in then that will also be considered in the
+    final solution. Following example is that type.
```
Rewrite the above statement, check grammar and language.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+
+    * Non linear system having non polynomial equation, if there is
+    complex solution for particular symbol and no real solution, then
+    `nonlinsolve` returns it's general solution.
```
[Same here](https://github.com/sympy/sympy/pull/11111/files#r67553885)

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    * Non linear system having all the equations polynomial, then it
+    returns both real and complex solutions.
+
+    Example
```
You need to write this only once.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    * If system is linear positive dimensional system, `nonlinsolve` can
+    solve this system also, since it is using `groebner` method. But it
+    is not recommended to solve linear system using `nonlinsolve`, because
+    `linsolve` is better for all kind of linear system.
```
This statement can also be rewritten to make more sense.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    {(3*z - 5, -z + 4, z)}
+
+
+    More examples:
```
Check out other modules to see, how example are written in SymPy.

Shekharrajak replied:
Thanks, I updated it.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    ========
+
+    >>> from sympy.core.symbol import symbols
+    >>> from sympy import pprint
```
Why is this required?

Shekharrajak replied:
I forgot to remove `pprint`. sorry.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    Note :
+    =======
+
+    1. `Dict` is defined in `SymPy`. `Finiteset` can't be used with
```
What's the issue here?

Shekharrajak replied:
We can't do `FiniteSet({key: value})`. But we can `FiniteSet(Dict({key: value}))`.

aktech replied:
Why do you need to put `dict` inside `FiniteSet`?

Shekharrajak replied:
I was using `FiniteSet` from the starting. Should I change it to `list` ?

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    Python `dict`.
+
+    """
+    from sympy.solvers.solvers import _invert as _invert_solver
```
Why are you using old `_invert`?

Shekharrajak replied:
solver's `_invert` return tuple (i, d) where ``i`` is independent of ``symbols`` and ``d``
contains symbols. (symbols means more than one variable, `solveset` _invert is for one variable.)

aktech commented:
Few comments:

* Use pep8 (checkout plugin for your editor).
* Heavily document your code, more documentation doesn't hurts. Most of the solvers is heuristics, there isn't much reference to the code we write. Others reading your code may not have the same understanding/mindset as yours, so its very important to explicitly mention, what you are doing & why you are doing. It's also very important for the maintainability of code.
* Break large blocks of logic into functions.
* Explain, why your multivariate solver doesn't misses a solution or why its better than old one, or why its upto mark based on the principles for `solveset`.
* Use proper variable names, explicit is better than implicit. `unsolved_equation` is a better variable name than `ue`.
* Don't make assumption based on simple cases, make sure the code works for general cases. [For e.g.](https://github.com/sympy/sympy/pull/11111#discussion_r67551590) [This is very important!].
* Re-read all your documentation from the perspective of a developer who is looking at your code for the first time and see if it make sense, without much background information.

Shekharrajak commented:
Thanks @aktech , I hope, I have resolved these mistakes in new commit.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,525 @@ def linsolve(system, *symbols):
+                if isinstance(soln, Complement):
+                    # extract solution and complement
+                    complements[s] = soln.args[1]
+                    soln = soln.args[0]
```
See my point 6 again. I don't understand the logic here. Please don't just assume and write anything, make sure you can justify, what you write.

I have already told you, don't write for particular case, It may not work for all cases!

```python
In [14]: soln = Complement(ImageSet(Lambda(x, x**2), N), Interval(0, oo))
In [15]: soln.args[0]
Out[15]: ImageSet(Lambda(x, x**2), <function N at 0x1045e6b70>)
```

Shekharrajak replied:
```
In [5]: soln_imageset
Out[5]: 
⎧ 2                                     ⎫         
⎨x  | x ∊ <function N at 0x7f89e7028b18>⎬ \ [0, ∞)
⎩                                       ⎭         

In [6]: soln = Complement(ImageSet(Lambda(x, x**2), N), Interval(0, oo))

In [7]: soln = soln.args[0]

In [8]: soln
Out[8]: 
⎧ 2                                     ⎫
⎨x  | x ∊ <function N at 0x7f89e7028b18>⎬
⎩                                       ⎭

# then in line 1233, 1234

In [9]: soln_imageset = soln

In [10]: soln = FiniteSet(soln.lamda.expr)

In [11]: soln
Out[11]: 
⎧ 2⎫
⎨x ⎬
⎩  ⎭

```
`x**2` will be used in further steps.
At the end we include `soln_imageset`.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1117,3 +1120,525 @@ def linsolve(system, *symbols):
+                elif isinstance(soln, Intersection):
+                    # sometimes solveset returns Intersection with S.Real
+                    # when #11174 is fixed,then this line should be removed
+                    soln = soln.args[1]
```
Same here.

Shekharrajak replied:
Because of this issue https://github.com/sympy/sympy/issues/11174 I added these lines. 
eg.
```
In [2]: solveset(x + y - 2, x ,S.Reals)
Out[2]: (-∞, ∞) ∩ {y}

In [3]: x, y, z = symbols('x, y, z', real = True)

In [4]: solveset(x + y - 2, x ,S.Reals)
Out[4]: {-y + 2}
```
So I didn't edited this because of the issue. 
My actual intention was if soln is like `Intersection( FiniteSet(-y), Interval(-oo, 0))`

```
In [25]:  Intersection(Interval(-oo, 0), FiniteSet(-y))
Out[25]: [0, ∞) ∩ {y}

In [26]:  Intersection( FiniteSet(-y), Interval(-oo, 0))
Out[26]: [0, ∞) ∩ {y}

In [27]:  Intersection( FiniteSet(-y), Interval(-oo, 0)).args[0]
Out[27]: [0, ∞)

In [28]:  Intersection( FiniteSet(-y), Interval(-oo, 0)).args[1]
Out[28]: {y} # choose this for further steps

```
and in final soln add with intersection.

I will change it.

aktech commented:
> When I tried to add both complex and real solution. I found that complex solution comes complicated which makes the further steps hard.

What do you mean by hard? The point is to return correct solution to the user, not just to make the computation easier or hard. Please justify, what's your understanding here.

> Also sometimes solveset returns same values for both real and complex domain.

So, what?

> So for now if no real solution then only using complex solution. (means either real or complex solution).

I don't understand, what do you want to achieve by that.

Shekharrajak committed:
both real and complex soln from substitution method

Shekharrajak commented:
@aktech , I have modified the code to get both real and complex solution. Now I am using [`_solve_using_know_values`](https://github.com/sympy/sympy/pull/11111/commits/a984c90d98efd6acf13c0c69bab5913f17167aeb#diff-eec0422923e8f100745c015cd8fdd6cfR1204) helper function. Doing `solveset_real` and then `solveset_complex` then combining both soln. 
Thanks for comments.

Shekharrajak committed:
more testcases added.

Shekharrajak commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -1012,6 +1012,251 @@ def test_linsolve():
+def test_issue_5132():
+    x, y, z, r, t = symbols('x, y, z, r, t', real=True)
+    # check : both soln seems equal
+    # wolfram -s_x,-s_y and +s_x,+s_y without Abs
```
some comments are temporary in test file.

Shekharrajak committed:
substitution returns all(real and complex) solns

Shekharrajak committed:
substitution returns all(real and complex) solns

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,631 @@ def linsolve(system, *symbols):
+def substitution(system, symbols=None, result=[{}], known_symbols=[], exclude=[],
+                 all_symbols=None):
+    r""" Solves the `system` using substitution method.
+    A helper function for     `nonlinsolve`. This will be called from
```
why so much space between `for` and `nonlinsolve`?
Please fix pep8.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,631 @@ def linsolve(system, *symbols):
+                 all_symbols=None):
+    r""" Solves the `system` using substitution method.
+    A helper function for     `nonlinsolve`. This will be called from
+    `nonlinsolve` when any equation(s)     is non polynomial equation.
```
same here

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,631 @@ def linsolve(system, *symbols):
+    >>> from sympy.core.symbol import symbols
+    >>> x, y = symbols('x, y', real = True)
+    >>> from sympy.solvers.solveset import substitution
+    >>> substitution([x +y], [x], [{y: 1}], [y], set([]), [x, y])
```
space around operator!

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,631 @@ def linsolve(system, *symbols):
+
+    * when you want soln should not satisfy eq `x + 1 = 0`
+
+    >>> substitution([x +y], [x], [{y: 1}], [y], set([x + 1]), [y, x])
```
[same here](https://github.com/sympy/sympy/pull/11111/files#r68220056)

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,631 @@ def linsolve(system, *symbols):
+
+    >>> substitution([x +y], [x], [{y: 1}], [y], set([x + 1]), [y, x])
+    EmptySet()
+    >>> substitution([x +y], [x], [{y: 1}], [y], set([x - 1]), [y, x])
```
[same here](https://github.com/sympy/sympy/pull/11111/files#r68220056)

Shekharrajak replied:
Thank, I will fix them in next commit with changes in `nonlinsolve`.

Shekharrajak committed:
better substitution, nonlinsolve when some eq is poly and some nonpolys

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,625 @@ def linsolve(system, *symbols):
+    >>> foo =  a + b + c + d
+    >>> bar = a*b + b*c + c*d + d*a
+    >>> foo_bar = a*b*c + b*c*d + c*d*a + d*a*b
+    >>> bar_foo = a*b*c*d -1
```
Better names for the variables.

Shekharrajak replied:
Edited. Thanks

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,625 @@ def linsolve(system, *symbols):
+    >>> system = [foo, bar, foo_bar, bar_foo]
+    >>> is_zero_dimensional(system)
+    False
+    >>> pprint(nonlinsolve(system, [a, b, c, d]), use_unicode=False)
```
The API seems to be counter-intuitive to me.
Don't you agree that all the values should have the same type?

Shekharrajak replied:
Not sure.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,625 @@ def linsolve(system, *symbols):
+    >>> nonlinsolve([(x+y)**2 - 4, x + y - 2], [x, y])
+    {(-y + 2, y)}
+
+    * If Non linear system having non polynomial equation have complex
```
Are you returning a combined set of real and complex solutions?
I don't seem to understand the point you are trying to make here.
Can you kindly elaborate on this ?

Shekharrajak replied:
I have edited the sentence.

"If some of the equations are non polynomial eq then `nonlinsolve`
    will call `substitution` function and returns real and complex solutions ."

Shekharrajak replied:
I have edited the sentence.

"If some of the equations are non polynomial eq then `nonlinsolve`
    will call `substitution` function and returns real and complex solutions ."

kshitij10496 replied:
That is not helping !

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,625 @@ def linsolve(system, *symbols):
+    returns it's general solution.
+
+    >>> from sympy import exp, sin
+    >>> nonlinsolve([exp(x) - sin(y), y**2 - 4], [x, y])
```
I think the output should be based on some specified `domain` of the symbols.
For  `ℝ^2`, the result should be `{(log(sin(2)), 2)}` 
While for the complex regions, it should be the last 2 solutions.
Ping @aktech @hargup 
What are your thoughts on this ?

kshitij10496 replied:
This issue has not been addressed yet.
I don't know why is it showing as an outdated diff !

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,625 @@ def linsolve(system, *symbols):
+        Mod(log(sin(2)), 2*I*pi)), Integers()), 2)}
+
+
+    * If Non linear system have all the equations polynomial, then it
```
Do you mean that the input is a system of polynomial equations ?

Shekharrajak replied:
Yes, I will edit this sentence. Thanks.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,625 @@ def linsolve(system, *symbols):
+    {(-2, -1), (2, 1), (-sqrt(2)*I, sqrt(2)*I), (sqrt(2)*I, -sqrt(2)*I)}
+
+
+    * If system is linear and positive dimensional system, `nonlinsolve` can
```
I think this refers to a system of over-determined linear equations.
This has already been implemented in `linsolve`.
Why do we have this feature here as well ?

Shekharrajak replied:
"But it is not recommended to solve linear system using `nonlinsolve`,
    because `linsolve` is better for all kind of linear system."
I wrote there.

kshitij10496 replied:
> "But it is not recommended to solve linear system using nonlinsolve,
because linsolve is better for all kind of linear system."

I think simply calling `linsolve` should suffice.

So why do we have to rewrite this code ?
Is it more efficient than `linsolve` ?

Shekharrajak replied:
>So why do we have to rewrite this code ?

see the previous line : 
"
`nonlinsolve` can solve some linear(zero or positive dimensional)
    system (because it is using `groebner` function to get the
    groebner basis and then `substitution` function basis as the new `system`)
"

>Is it more efficient than linsolve ?

see last line

" `linsolve` is better for all kind of linear system."

I think we should write more things about functions that it can do and can't do.
 
Please comment on new commit rather than old.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,625 @@ def linsolve(system, *symbols):
+    How nonlinsolve is better than old solver `_solve_system` :
+    ===========================================================
+
+    * A positive dimensional system solver : nonlinsolve can return
```
Are we certain of solving all the positive dimensional systems via `nonlinsolve`?

Shekharrajak replied:
I hope it can. If you know any +ve dimensional system examples then please let me know.

kshitij10496 replied:
>  I hope it can.

I think we shouldn't generalise that `nonlinsolve` can solve a positive dimensional system based on a single example. I am pretty sure that simple googling would result in several examples of a positive dimensional system.

Shekharrajak replied:
some are [here](https://github.com/sympy/sympy/pull/11111/files/573acf5b91154ff2efa587de1557658912140ec8#diff-85baa04bbf4e1dfd9128782738e45424R1029).

Shekharrajak committed:
using unrad to remove radicals mostly sqrt

Shekharrajak commented:
@aktech now it is free from solvers `_invert`. Please check once.

Shekharrajak committed:
minor changes

minor change *symbols

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -509,6 +512,8 @@ def _solve_radical(f, symbol, solveset_solver):
         result = Union(*[imageset(Lambda(y, g_y), f_y_sols)
                          for g_y in g_y_s])
 
+    if isinstance(result, ConditionSet):
```
Can you add a testcase corresponding to this clause ?

Shekharrajak replied:
Okay I will do in next commit.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,597 @@ def linsolve(system, *symbols):
+    Returns
+    =======
+
+    A FiniteSet of ordered tuple of values of `symbols` for which the `system`
```
A tuple is always ordered so what do you mean by this statement ?

Shekharrajak replied:
I followed [these word](https://github.com/Shekharrajak/sympy/blob/16496485e04c21b893916c3449471c199c79be6f/sympy/solvers/solveset.py#L979)

kshitij10496 replied:
Since you copied the statements from `linsolve`, can you justify their relevance here ?

Shekharrajak replied:
>justify their relevance

Output API is same for both `FniteSet(tuple_1, tuple_2,...tuple_n)`.

kshitij10496 replied:
>  Output API is same for both FniteSet(tuple_1, tuple_2,...tuple_n).

`{(-1/d, d, 1/d ,{d} \ {0})}`
We should specify this kind out output API as well.

hargup replied:
> I followed these word

@Shekharrajak: Just because something is written somewhere doesn't mean it is correct, please use your own judgment to justify the code you write.

Though saying `ordered tuple` is redundant, it might add clarity for users who are not very well versed with python. So I'll say keep it.

Shekharrajak replied:
Okay I will not do this again.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,597 @@ def linsolve(system, *symbols):
+    False
+    >>> pprint(nonlinsolve(system, [a, b, c, d]), use_unicode=False)
+      -1       1
+    {(---, -d, -, {d} \ {0})}
```
This makes the output inconsistent, doesn't it ?
Ping @aktech @hargup

hargup replied:
Yes, `{d}\{0}` is kinda inconsistent, but I see no way around it. So lets rather keep it, until we find something better.

hargup replied:
That something better need not block this PR.

kshitij10496 replied:
>  Yes, {d}\{0} is kinda inconsistent, but I see no way around it. So lets rather keep it, until we find something better.

That's cool.
This ought to be documented for future reference, shouldn't it ?

hargup commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1126,625 @@ def linsolve(system, *symbols):
+def nonlinsolve(system, symbols):
+    r"""
+    Solve system of N non linear equations with M variables, which means both
+    under and overdetermined systems are supported. Positive dimensional
```
Yes, an additional line of comment will do.

aktech commented:
@Shekharrajak Does the tests includes 100% of the old solve's multivariate solver?

Shekharrajak commented:
@aktech Yes. I am sure.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1124,597 @@ def linsolve(system, *symbols):
+    significance except for the fact it is just used to maintain a consistent
+    output format throughout the solveset.
+
+    For the given set of Equations, the respective input types
```
The examples serve the purpose of familiarising the user with the API.
So, I find this unnecessary .

Instead of this, it will be helpful to state the circumstances under which various exceptions are raised by this function.

Shekharrajak replied:
I think both should be present. I added exceptions that ` nonlinsolve` raises as well.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1124,597 @@ def linsolve(system, *symbols):
+        Mod(log(sin(2)), 2*I*pi)), Integers()), 2)}
+
+
+    * If system is Non linear polynomial zero dimensional then it returns
```
Do you mean a zero-dimensional polynomial system ?

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1124,597 @@ def linsolve(system, *symbols):
+    >>> nonlinsolve([x + 2*y -z - 3, x - y - 4*z + 9 , y + z - 4], [x, y, z])
+    {(3*z - 5, -z + 4, z)}
+
+    * system having polynomial equations:
```
Why have you written this case separately ?
Also, the example given below is a zero-dimensional system.

Shekharrajak replied:
>Why have you written this case separately ?

I wrote all the cases separately.

>Also, the example given below is a zero-dimensional system.

`nonlinsolve` can solve both zero dimensional and positive dimensional system.

kshitij10496 replied:
> I wrote all the cases separately.

Why are we writing two cases for solving a zero-dimensional system ?
Is there something else in this example that you want to specify ?

Shekharrajak replied:
Here only real solution is present. I will add that line.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1124,597 @@ def linsolve(system, *symbols):
+        raise ValueError('Symbols must be given, for which solution of the '
+                         'system is to be found.')
+
+    if hasattr(symbols[0], '__iter__'):
```
What are you trying to check by this condition ?

Shekharrajak replied:
`nonlinsolve(system, *symbols)` .So my intention is to take symbols[0]. If `symbols[0]` have `iterable` property(list or something like that). But in `substitution it is not needed(In `substitution` function we have more parameters) I have removed this `if ` block there.
Thanks.

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1124,597 @@ def linsolve(system, *symbols):
+        sym = symbols[0].is_Symbol
+    except AttributeError:
+        sym = False
+
```
```python
In []: nonlinsolve(system, [a, pi]) # using the positive-dimensional system you have defined above in the docs.
Out[]: {(-b - c - d, pi)}
```

Shekharrajak replied:
The example is in `nonlinsolve` docs (below). If only `[a, pi]` is passed as `symbols` then it will take other symbols(b, c, d) as const (user is not interested for these symbols) and value of ` pi` is `pi` .
Should we solve for all the variables present in the equations ?

kshitij10496 commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1124,597 @@ def linsolve(system, *symbols):
+    if result:
+        return result
+    else:
+        return S.EmptySet
```
Are we certain that the system of equations is inconsistent when this clause is activated? 
If not, why are we returning an `EmptySet`?

Shekharrajak replied:
Right now there is two cases [test1](https://github.com/sympy/sympy/pull/11111/files#diff-85baa04bbf4e1dfd9128782738e45424R1041), [test2](https://github.com/sympy/sympy/pull/11111/files#diff-85baa04bbf4e1dfd9128782738e45424R1224) I know when it returns EmptySet and it is correct also.

It doesn't check whether system is inconsistent or not . But `solveset_real`  ( or `solveset_complex` used in `substitution`) will return solution for at least some variables if there is soln. If `solveset_real`  (or or `solveset_complex` used in `substitution`) return `S.EmptySet` for all the variables present in symbols then it will return `S.EmptySet`. If `solveset` returns `ConditionSet` for all the variables then at that time `S.EmptySet` looks incorrect. What should it return in that case? I think `solveset` will solve for at least some variable and will not return `conditionset` all the time in `substitution` function.

Shekharrajak committed:
conditionset return, when solveset failed to solve all the eqs

minor changes

minor change

test issue 5132_2 fixed

aktech commented:
@Shekharrajak Please see point 3 here: https://github.com/sympy/sympy/pull/11111#issuecomment-226851817

Shekharrajak commented:
@aktech , yes I am trying to break the function [_solve_using_known_values](https://github.com/sympy/sympy/pull/11111/commits/2ded376e9e969519bae3a3142b8993767e452d09#diff-eec0422923e8f100745c015cd8fdd6cfR1273) into few helper function . Right now I have defined 2 helper functions : [_extract_main_soln](https://github.com/sympy/sympy/pull/11111/commits/2ded376e9e969519bae3a3142b8993767e452d09#diff-eec0422923e8f100745c015cd8fdd6cfR1284) , [_append_new_soln](https://github.com/sympy/sympy/pull/11111/commits/2ded376e9e969519bae3a3142b8993767e452d09#diff-eec0422923e8f100745c015cd8fdd6cfR1410)

Shekharrajak commented:
One cases is [failing](https://travis-ci.org/sympy/sympy/jobs/142046403#L3818) because of `checksol` bug. I am trying to fix that bug in this PR https://github.com/sympy/sympy/pull/11339

Shekharrajak committed:
break into functions

extract_main_soln

hit variable not needed, removed

extract_main_soln returns finiteset

_append_new_soln new func

some new testcases

Shekharrajak committed:
trig xfail added

minor changes

Shekharrajak committed:
doctest error fixed

return infinite soln when we have certain soln and general form in result list

symbol sort before groebner

2 soln for doctest positive dim system

aktech commented:
@Shekharrajak are all the comments mentioned here: https://github.com/sympy/sympy/pull/11111#issuecomment-226851817 addressed?

Shekharrajak commented:
@aktech , Yes.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1125,823 @@ def linsolve(system, *symbols):
+    eqs_in_better_order = list(
+        ordered(system, lambda _: len(_unsolved_syms(_))))
+
+    def _return_conditionset():
```
There is no point in defining a function which uses the local variable of another function as it would not be useful outside the scope.

aktech commented:
You should not define a function inside a function, unless you really have too. The point of the statement `Break large blocks of logic into functions.` is to separate out chunks of logic into "meaningful functions" so as to make it reusable.

Shekharrajak committed:
helper functions edited

Shekharrajak commented:
@aktech , I have updated the PR ([commit](https://github.com/sympy/sympy/pull/11111/commits/386a3649527d2451f4e8362457c510b9bfcd3eae)). Now no function is defined inside helper function. Thanks for review.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1125,831 @@ def linsolve(system, *symbols):
+    from sympy.polys.polytools import groebner
+    # substitution method where new system is groebner basis of the system
+    _symbols = symbols
+    _symbols.sort(key=default_sort_key)
```
The `tuple` object has no attribute `sort`.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1125,831 @@ def linsolve(system, *symbols):
+        Values are known for these variable(s)
+    result : An empty list or list of dict
+        If No symbol values is known then empty list otherwise
+        symbol as keys and corresponding value in dict.
```
It's FiniteSet of tuple.

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1121,3 +1125,831 @@ def linsolve(system, *symbols):
+
+def substitution(system, symbols, result=[{}], known_symbols=[],
+                 exclude=[], all_symbols=None):
+    r""" Solves the `system` using substitution method.
```
Use 79 characters in a line.

aktech commented:
@Shekharrajak `nonlinsolve` fails for some of the very trivial cases.

```python
In [29]: nonlinsolve([x-y, 0], x, y)
[...]
AttributeError: 'int' object has no attribute 'as_numer_denom'
```

```python
In [30]: nonlinsolve([x-y, x-y], x, y)
[....]
AttributeError: 'tuple' object has no attribute 'sort'
```

```python
In [51]: nonlinsolve([x, 0], x, y)
[....]
AttributeError: 'int' object has no attribute 'as_numer_denom'
```

```python
In [56]: nonlinsolve([Abs(x) - y], x, y)

ValueError:
Absolute values cannot be inverted in the complex domain.
```

A major chunk of the code is untested (checkout `bin/coverage_report.py`), the coverage for any new code should 100% except for the trivial lines.

> @aktech , I have updated the PR (commit). Now no function is defined inside helper function. Thanks for review.

Let me know the improvements we achieved by that.

Shekharrajak commented:
Thanks @aktech , I am trying to resolve these issues.

Shekharrajak committed:
minor edit

Shekharrajak committed:
more testcases added and few changes

jksuom commented on sympy/solvers/solveset.py:
```diff
@@ -1192,3 +1196,837 @@ def linsolve(system, *symbols):
+                eq = eq_unrad
+        if isinstance(eq, Expr):
+            eq = eq.as_numer_denom()[0]
+            # this will remove sqrt if symbols are under it
```
This will remove *numeric* roots from the generators but not those that involve symbols.

Shekharrajak replied:
Thanks , actually that comment is not needed.

jksuom commented on sympy/solvers/solveset.py:
```diff
@@ -1192,3 +1196,837 @@ def linsolve(system, *symbols):
+            eq = eq.as_numer_denom()[0]
+            # this will remove sqrt if symbols are under it
+            poly = eq.as_poly(*symbols, extension=True)
+        if poly is not None:
```
Is `poly` defined only on the previous line? What if `eq` is not an expression?

Shekharrajak replied:
I think it should check for `Number` also, if `eq` is a Number then eq will not be added in `poly` and `nonpoly` list ( since that will not help for finding soln). It seems I forgot to add line `poly = None` at the starting. I have edited this in new commit. Thanks.

Shekharrajak commented:
The coverage report link : https://shekharrajak.github.io/sympy_11111_coverage_report/sympy_solvers_solveset_py.html#n1214

Lines [1378](https://shekharrajak.github.io/sympy_11111_coverage_report/sympy_solvers_solveset_py.html#n1378), 1626, 1739 are for adding intersection in final solution. I don't find the testcase for that, but it is similar to adding complement (line 1382), so it will work correctly.

Line [2023](https://shekharrajak.github.io/sympy_11111_coverage_report/sympy_solvers_solveset_py.html#n2023), `_solve_poly_system` works well(I don't have testcase, when it fails). If it fails then `substitution` can handle that system(I have checked it manually). Similarly line [1634](https://shekharrajak.github.io/sympy_11111_coverage_report/sympy_solvers_solveset_py.html#n1634).

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1309,22 +1309,24 @@ def substitution(system, symbols, result=[{}], known_symbols=[],
     if not symbols:
-        raise ValueError('Symbols must be given, for which solution of the '
-                         'system is to be found.')
+        msg = 'Symbols must be given, for which solution of the '\
```
Simply do this:
```python
msg = ('Symbols must be given, for which solution of the '
       'system is to be found.')
```

Avoid using `\`

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1309,22 +1309,24 @@ def substitution(system, symbols, result=[{}], known_symbols=[],
+        raise ValueError(filldedent(msg))
+
+    if not is_sequence(symbols):
+        msg = 'symbols should be given as a sequence, e.g. a list.' \
```
same here

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1309,22 +1309,24 @@ def substitution(system, symbols, result=[{}], known_symbols=[],
-    if not is_sequence(symbols):
-                raise TypeError(
-                    'syms should be given as a sequence, e.g. a list')
+        msg = 'Iterable of symbols must be given as' \
```
same here

aktech commented on sympy/solvers/solveset.py:
```diff
@@ -1987,13 +1994,14 @@ def nonlinsolve(system, *symbols):
-        raise ValueError('Symbols or iterable of symbols must be given as '
-                         'second argument, not type \
-            %s: %s' % (type(symbols[0]), symbols[0]))
+        msg = 'Symbols or iterable of symbols must be given as \
```
same

aktech commented:
Apart from the couple of minor comments, It looks fine to me. We can push it for now and improve later. @Shekharrajak Thanks for your work.

Will push in 24hr after the minor comments are addressed, if there is no objection.

Shekharrajak committed:
minor changes

missing space added

Shekharrajak commented:
Thanks for review.

aktech commented:
@Shekharrajak Thanks for working on it. Its in.

Shekharrajak commented:
Thanks.
